Documentation for HSM buttons (#269)
hellohaptik/docs_haptik: docs/whatsapp/sending-notifications.md

```

Note: The above structure is for reference only and will change based on the way we create HSM on Whatsapp Business Manager and for each media type.

> Only Images, PDF, MP3 and MP4 files are supported.

## Adding buttons to HSM templates

> WhatsApp allows buttons to be added to the HSM template. The `Buttons` section is located at the bottom of the HSM page (assuming that you are either creating a new HSM or editing an existing one). There are two broad categories of buttons that you can add to your template:

Reviewer: Can we add an image here?

Author: bro, there are images added.. its just that since this is a PR on github, they are not visible. Once they are launched we will be able to see them.

1. Quick Reply
2. Call To Action

**Quick Reply:** <br/>

> A HSM template can have at the most 3 quick reply buttons. Quick Reply buttons as their name implies, are buttons which when tapped-on by the user, send the title of those buttons as a message from that user. _(You can even provide a payload to override the message that gets sent back on button click. More on that later, in sending HSM Templates section)_

Reviewer 2: ```suggestion
An HSM template can have at the most 3 quick reply buttons. Quick Reply buttons as their name implies, are buttons which when tapped-on by the user, send the title of those buttons as a message from that user. _(You can even provide a payload to override the message that gets sent back on button click. More on that later, in sending HSM Templates section)_
```

> A HSM template can have at the most 3 quick reply buttons. Quick Reply buttons as their name implies, are buttons which when tapped-on by the user, send the title of those buttons as a message from that user. _(You can even provide a payload to override the message that gets sent back on button click. More on that later, in sending HSM Templates section)_

Reviewer 2: ```suggestion
A HSM template can have at the most 3 quick reply buttons. Quick Reply buttons as their name implies, are buttons which when tapped-on by the user, send the title of those buttons as a message from that user. _(You can even provide a payload to override the message that gets sent on button click. More on that later, in sending HSM Templates section)_
```

    - When `Call Phone Number` is selected in `Type of Action` you get the following columns adjacent to it viz. `Button Text`, `Country`, `Phone Number`.

- _Visit Website_:
    - This button takes a website URL and text to display on the button as parameters and when sent to the user, this button on click will actually launch the website in the user's phone.
    - When `Visit Website` is selected in `Type of Action` you get the following columns adjacent to it viz. `Button Text`, `URL Type`, `Website URL`.
    - `URL Type` is a drop down in itself which has two options viz. `Static` and `Dynamic`
    - A `Static` URL is self explanatory. It is basically a hardcoded website URL.
    - A `Dynamic` URL is a URL that will have a parameter which will be suffixed to it to form the final URL. <br/> For example, if the URL entered for Dynamic field is `https://<sitedomain>.com`, then while sending HSM, a paramter will have to be sent which will be suffixed to this URL. Thus, if you send the parameter as `1234` then when the user receives this HSM and clicks on this button, the URL launched would be `https://<sitedomain>.com/1234`.

>     - Do note that sending the parameter for `Dynamic` URL in the HSM sending request is mandatory.

Reviewer 2: ```suggestion
    - Do note that sending the parameter for `Dynamic` URL in the HSM while sending request is mandatory.
```

- ![Call To Action buttons in HSM template](assets/hsm_buttons/call-to-action-button.png) <br/> _Call To Action buttons in HSM template builder_

- ![Call To Action buttons in user's WhatsApp Messenger](assets/hsm_buttons/whatsapp_call-to-action.jpg) <br/> _Call To Action buttons in user's WhatsApp Messenger_

## Sending HSMs with Buttons

[Link to the official documentation.](https://developers.facebook.com/docs/whatsapp/api/messages/message-templates/interactive-message-templates)

**Quick Reply:**

> 

Reviewer: Is there a limit to the QR, button text? Character limit

Author: excellent point. there indeed is a character limit.. have upgraded the document to include that

    - For instance, in our above example we have a template which sends two Quick Reply buttons viz, `Yes` and `No`.
    - Let's say when the HSM is sent to the user, the button `Yes` is configured without any payload, but the button `No` has a _payload_ `No, I did not receive the order`.
    - Now, inside the WhatsApp Messenger, if the user clicks on `Yes`, the reply sent to Haptik API will be `Yes`.
    - However, if the user clicks on `No`, the reply sent back to Haptik API will be `No, I did not receive the order` i.e. the payload will be sent back.

Sample CURL request for sending payload for Quick Reply. _(Note how index is used to target specific buttons)_

```
curl -X POST <base_url>/whatsapp/notification/v2/ \
-H 'client-id: <client-id>' \
-H 'Authorization: Bearer <token>' \
-H 'Content-Type: application/json' \
-d '{
  "business_id": <ID of the business>,
  "to": "<phone number>",
  "type": "template",
  "template": {
    "namespace": "<from whatsapp dashboard>",
    "name": "<name of the template>",
    "language": {
      "policy": "deterministic",
      "code": "en"
    },
    "components": [
      {
        "type": "button",

>         "sub_type": "quick_reply",

Reviewer 3: Should we explain what `sub_type` is?

Author: I don't feel so.. They would never need to change the values inside the `sub_type`.
Do correct me if I am wrong

Reviewer 3: Sounds alright 🧐

>         "index": "1",

Reviewer 3: Do we need to explain what this `index` is?

Author: Very good point.. made the changes..

Reviewer 2: Its better we explain this @RandiveSushant

        "parameters": [{
          "type": "payload",
          "payload": "No, I did not receive the order"
        }]
      }
    ]
  }
}
'
```

**Call To Action:**

- Sending a HSM template with Call To Action buttons is similar to sending any other HSM template.
- The only difference is when you would like to send a parameter to the dynamic URL.

> - Do note that the paramter sent to the dynamic URL should be a URL safe string, since it would be appended to the end of the base_URL provided in the HSM builder template _(Please refer to the earlier secitons for detailed explanation)_.

Reviewer 2: ```suggestion
- Do note that the paramter sent to the dynamic URL should be a URL safe string, since it would be appended at the end of the base_URL provided in the HSM builder template _(Please refer to the earlier sections for detailed explanation)_.
```

Sample CURL request for sending parameter for dynamic URL. _(Note how index is used to target the dynamic URL button)_

```
curl -X POST <base_url>/whatsapp/notification/v2/ \
-H 'client-id: <client-id>' \
-H 'Authorization: Bearer <token>' \
-H 'Content-Type: application/json' \
-d '{
  "business_id": <ID of the business>,
  "to": "<phone number>",
  "type": "template",
  "template": {
    "namespace": "<from whatsapp dashboard>",
    "name": "<name of the template>",
    "language": {
      "policy": "deterministic",
      "code": "en"
    },
    "components": [
      {
        "type": "button",
        "sub_type": "url",
        "index": "1",
        "parameters": [{

>           "type": "text",

Reviewer 2: Where does this **type** and **text** get used up?

Author: So, this is how WhatsApp asks you to specify the type of parameter you are sending.. 
did not dive too deeply in explaining this in the doc since the our clients would seldom need to touch the value here

Reviewer 2: Alright... makes sense
